modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryCancelTests.java

```diff
@@ -0,0 +1,182 @@
+
+        // Waits for the indexing operation listener to block
+        barrier.await(30, TimeUnit.SECONDS);
+
+        // Status should show running
+        ListTasksResponse tasksList = client().admin().cluster().prepareListTasks()
+                .setActions(DeleteByQueryAction.NAME).setDetailed(true).get();
+        assertThat(tasksList.getNodeFailures(), empty());
+        assertThat(tasksList.getTaskFailures(), empty());
+        assertThat(tasksList.getTasks(), hasSize(1));
+        BulkByScrollTask.Status status = (BulkByScrollTask.Status) tasksList.getTasks().get(0).getStatus();
+        assertNull(status.getReasonCancelled());
+
+        // Cancel the request while the deletions are blocked. This will prevent further deletions requests from being sent.
+        List<TaskInfo> cancelledTasks = client().admin().cluster().prepareCancelTasks()
+                .setActions(DeleteByQueryAction.NAME).get().getTasks();
+        assertThat(cancelledTasks, hasSize(1));
+
+        // The status should now show canceled. The request will still be in the list because the script is still blocked.
+        tasksList = client().admin().cluster().prepareListTasks().setActions(DeleteByQueryAction.NAME).setDetailed(true).get();
+        assertThat(tasksList.getNodeFailures(), empty());
+        assertThat(tasksList.getTaskFailures(), empty());
+        assertThat(tasksList.getTasks(), hasSize(1));
+        status = (BulkByScrollTask.Status) tasksList.getTasks().get(0).getStatus();
+        assertEquals(CancelTasksRequest.DEFAULT_REASON, status.getReasonCancelled());
+
+        // Now unblock the listener so that it can proceed
+        barrier.await();
+
+        // And check the status of the response
+        BulkIndexByScrollResponse response = future.get();
+        assertThat(response, matcher()
+                .deleted(lessThanOrEqualTo((long) MAX_DELETIONS)).batches(MAX_DELETIONS).reasonCancelled(equalTo("by user request")));
+    }
+
+
+    public static class DeleteByQueryCancellationPlugin extends Plugin {
+
+        @Override
+        public String name() {
+            return "delete-by-query-cancellation";
+        }
+
+        @Override
+        public String description() {
+            return "See " + DeleteByQueryCancellationPlugin.class.getName();
+        }
+
+        @Override
+        public void onIndexModule(IndexModule indexModule) {
+            indexModule.addIndexOperationListener(new BlockingDeleteListener());
+        }
+    }
+
+    /**
+     * A {@link IndexingOperationListener} that allows a given number of documents to be deleted
+     * and then blocks until it is notified to proceed.
+     */
+    public static class BlockingDeleteListener implements IndexingOperationListener {
+
+        private final CountDown blockAfter = new CountDown(MAX_DELETIONS);
+
+        @Override
+        public Engine.Delete preDelete(Engine.Delete delete) {
```

Comment:
I like this a lot more than I like the way I do update-by-query and reindex cancellation. Do you think as a followup you can port those to this test?